### PR TITLE
Add Dictionary Deserialization Support and Improve Test Coverage for Edge Cases

### PR DESCRIPTION
**Pull request by joheredi**
### **Codegen Changes**

*   Added missing support for correct **Dictionary** deserialization.

### **Test Fixes**

*   **ArrayItem**: Fixed a false negative in `deepEqual` where `children: undefined` was treated as different from `children` not present. These should be considered equivalent.
*   **DictionaryDatetime**: Corrected comparison logic—previously converted dates to UTC string but compared against ISO format.
*   **Scalar Decimal and Decimal128 Verify**: Adjusted tests to use integer arithmetic due to known JS decimal precision limitations. Tracking JSDoc improvements in <https://github.com/Azure/autorest.typescript/issues/3611>.

### **Added Tests**

*   **Alternate ApiVersion tests**: Currently disabled, pending <https://github.com/Azure/autorest.typescript/issues/3304>.
*   **Conditional Request tests**: Verified handling of `304` and `412`. These are application-level semantics and left to customers. Documentation improvements tracked in <https://github.com/Azure/autorest.typescript/issues/3612>.
*   **Dictionary Recursive Model**
*   **Dictionary with Unknown Value**

### CODE REVIEW

File: packages/typespec-ts/test/azureModularIntegration/scalar.spec.ts [hunk 1]
```diff
@@ -98,30 +98,37 @@ describe("Scalar Client", () => {
...
+    const scaledSum = getResult.reduce(
+      (acc: number, val: number) => acc + Math.round(val * 100),
+      0
+    );
+    const total = scaledSum / 100;
+
+    await client.decimalVerify.verify(total);
   });
 
   it("should fail to post decimal128 verify", async () => {
```
Comment (MaryGao) [inline packages/typespec-ts/test/azureModularIntegration/scalar.spec.ts]:
```suggestion
  it("should pass the post decimal128 verify", async () => {
```

File: packages/typespec-ts/test/azureModularIntegration/scalar.spec.ts [hunk 2]
```diff
@@ -98,30 +98,37 @@ describe("Scalar Client", () => {
...
-        total += decimal;
-      });
-      await client.decimal128Verify.verify(total);
-      assert.fail("Expected an exception to be thrown.");
-    } catch (err) {
-      assert.strictEqual(JSON.parse(JSON.stringify(err)).statusCode, 400);
-    }
+    // JavaScript uses IEEE 754 floating-point, which cannot represent some decimal fractions exactly.
+    // Example: 0.1 + 0.1 + 0.1 === 0.30000000000000004
+    // To avoid precision errors when summing decimals, we convert to integer math (scale by 100).
```
Comment (MaryGao) [inline packages/typespec-ts/test/azureModularIntegration/scalar.spec.ts]: We have generated the doc for decimal: https://github.com/Azure/autorest.typescript/blob/d650bb58a945c2ce702af53031e82fbef98b7ea5/packages/typespec-ts/src/utils/modelUtils.ts#L1561-L1572. Maybe we could improve it there.

**Review by MaryGao — APPROVED** [review]
LGTM.

File: packages/typespec-ts/test/azureModularIntegration/arrayItemTypes.spec.ts [hunk 3]
```diff
@@ -40,10 +40,13 @@ const testedTypes: TypeDetail[] = [
...
     defaultValue: [1, "hello", null]
   },
-  // {
-  //   type: "model",
-  //   defaultValue: [{ property: "hello" }, { property: "world" }]
-  // },
+  {
+    type: "model",
+    defaultValue: [
+      { property: "hello", children: undefined },
```
[the remark this answers is not in the document]
Comment (joheredi) [reply packages/typespec-ts/test/azureModularIntegration/arrayItemTypes.spec.ts]: Yes, `{ property: "hello", children: undefined }` is the same as `{ property: "hello" }` but the test is being strict
